lib/rails/dom/testing/assertions/selector_assertions.rb

```diff
@@ -263,6 +263,35 @@ def assert_dom_email(&block)
         end
         alias_method :assert_select_email, :assert_dom_email
 
+
+        # Changes the depth of the document_root_element for the operations
+        # nested within the provided block. Accepts a singular Nokogiri node or
+        # an Nokogiri NodeSet as its argument.
+        #
+        #  within 'div[class~=?]', 'hello', count: 1 do
```

Comment:
This doesn't seem right, we shouldn't be hiding assertion calls in non-`assert` methods. If people want assertions, they should be using `assert_dom` directly.

Can we essentially do this?

```ruby
def within(selector_or_selection, &block)
  selector_or_selection = css_select(selector_or_selection) if selector_or_selection.is_a?(String)
  nest_selection nodeset(selector_or_selection), &block
end
```

Reply:
@kaspth good point.

I reached for `assert_dom` so that a missing element or a broken selector would result in a more helpful error message.

Otherwise, what should happen if the `within` arguments aren't on the page?

Reply:
@kaspth I've added some coverage to exercise what happens when the selector yields no results, or if the argument is nil.

I've removed the docs for supporting the results of `css_select` (handling a NodeSet). My thinking is that this method is best for moving the `document_root_element` to a single scope, but could be an alias for `assert_dom` when passed selectors and arguments.

Reply:
> I've added some coverage to exercise what happens when the selector yields no results, or if the argument is nil.

Cool, I like that!



> My thinking is that this method is best for moving the `document_root_element` to a single scope

I guess my main gripe is the passing of options like `count: 1`, which seems off for this. It feels like a detail that we haven't sweat and are passing onto users. I don't know within enough from Capybara, but do you ever expect to find a result with more than 1 match? We could bake a default `count: 1` in there (or is that `assert_dom`s default?).

I do like that we can substitutions like `text:`/`regex:`, or via the positional argument version.